js/ag_reference.js

```diff
@@ -0,0 +1,299 @@
+        return this._location;
+    },
+    
+    /**
+     * the displayname of the reference
+     * TODO not sure if this sis needed
+     * @type {object}
+     */
+    get displayname() {
+        //TODO
+        notimplemented();
+    },
+    
+    /**
+     * Visible columns
+     * @type {AttributeGroupColumn[]}
+     */
+    get columns () {
+        if (this._columns === undefined) {
+            var self = this;
+            this._columns = [];
+            
+            var addCol = function (col) {
+                if (col._visible) {
+                    self._columns.push(col);
+                }
+            };
+            
+            this._keyColumns.forEach(addCol);
+            this._aggregateColumns.forEach(addCol);
+            
+        }
+        return this._columns;
+    },
+    
+    /**
+     * The session object from the server
+     * @param {Object} session - the session object
+     */
+    /* jshint ignore:start */
+    set session(session) {
+        this._session = session;
+    },
+    /* jshint ignore:end */
+    
+    sort: function (sort) {
+        // TODO verify sort
+        var newLocation = this._location.sort(sort);
+        return new AttributeGroupReference(this._keyColumns, this._aggregateColumns, newLocation);
+    },
+    
+    search: function (term) {
+        // TODO verify term
+        var newLocation = this._location.search(term);
+        return new AttributeGroupReference(this._keyColumns, this._aggregateColumns, newLocation);
+    },
+    
+    read: function (limit) {
+        //TODO
+        /*
+        This is how we should create the attributegroup url:
+        <service>/catalog/<_catalogId>/attributegroup/<path>/<search>/<_keyColumns>;<_aggregateColumns><sort><page>
+         */
+        notimplemented();
+    }
+};
+
+function AttributeGroupPage(reference, data, hasPrevious, hasNext) {
+    this.reference = reference;
+    this.hasPrevious = hasPrevious;
+    this.hasNext = hasNext;
+    
+    //TODO use the data somehow
+}
+AttributeGroupPage.prototype = {
+    constructor: AttributeGroupPage,
+    
+    get tuples () {
+        //TODO
+        notimplemented();
+    },
+    
+    get next() {
+        //TODO same reference with the new after
+        notimplemented();
+    },
+    
+    get previous() {
+        //TODO same reference with the new before
+        notimplemented();
+    }
+};
+
+function AttributeGroupTuple(page, data) {
+    this.page = page;
+    
+    //TODO use the data
+}
+AttributeGroupTuple.prototype = {
+    constructor: AttributeGroupTuple,
+    
+    // TODO take a look at values comment.
+    get isHTML() {
+        if (this._isHTML === undefined) {
+            var value = this.values;
+        }
+        return this._isHTML;
+    },
+
+    get values() {
+        //TODO 
+        // returned value should be the same as current tuple.values 
+        // or we should make this (https://github.com/informatics-isi-edu/ermrestjs/issues/189) change now?
```

Comment:
I think we do want to move toward this approach with `tuple.values` being an array of objects like:
```
{
  value: <value>,
  isHtml: boolean,
  isMarkdown: boolean
}
```

Karl commented that we should add `isMarkdown`. I'm not sure whether or not he meant to replace `isHtml` with `isMarkdown` or if they should both be defined. I'm trying to think of a case where we define a value as HTML that is not also Markdown.